refactor: replace examples with focused rustdoc examples

kwsantiago opened the pull request:
Closes #1973

The PR is now scoped to the first step of the work discussed in #1973:

- Remove the unmaintained example binaries in examples/
- Remove the related workspace members from Cargo.toml
- Clean up docs and CI references that pointed to the removed examples

oleonardolima commented:
Thanks for working on this one.
It's failing on CI. Make sure to use the 'just' commands and try building the docs locally if that helps with debugging the failures.

kwsantiago commented:
@oleonardolima thank you. CI "should" be fixed now, at least it is locally. Can we try again?

oleonardolima commented:
Thanks, there are some unrelated already merged commits, you probably need to do a rebase on top of master. Also, please note that we follow the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for commit messages, which would also need an update.

kwsantiago commented:
@oleonardolima ok should be ready to test CI now. Thanks!

oleonardolima commented:
You'll need to perform another rebase to incorporate the CI updates/fixes, and remove the unrelated merge commit.

kwsantiago commented:
@oleonardolima should be good now, let me know if there is anything else here.

oleonardolima commented on Cargo.toml:
```diff
@@ -13,6 +13,11 @@ members = [
+[workspace.dependencies]
+tokio = "=1.38.1"
+reqwest = "=0.12.4"
+socket2 = "=0.5.10"
```
I don't think this is required, it should be already covered in CI by the `ci/pin-msrv.sh` used in CI.

oleonardolima reviewed (commented):
You should also remove the example exclusion in build and test CI step.

I'm also wondering if the `cfgs` are strictly required, but haven't tested anything different yet.

oleonardolima commented on crates/esplora/Cargo.toml:
```diff
@@ -24,7 +24,7 @@ futures = { version = "0.3.26", optional = true }
 bdk_chain = { path = "../chain" }
 bdk_testenv = { path = "../testenv" }
-tokio = { version = "1", features = ["rt", "rt-multi-thread", "macros"] }
+tokio = { workspace = true, features = ["rt", "rt-multi-thread", "macros"] }
```
You shouldn't need this too.

notmandatory commented:
cACK, but haven't done a close review yet.  Thanks for hanging in on this one, but looks like it needs another rebase.

oleonardolima commented:
I'm sorry for all the rebases you had to do, but you'll need another one to get the latest CI with bumped MSRV.

nymius commented on crates/chain/src/indexed_tx_graph.rs:
````diff
@@ -363,6 +372,21 @@ where
+    ///
+    /// let changeset = graph.apply_block_relevant(&block, 100);
+    /// ```
+>>>>>>> 8d6a3c64 (refactor: replace examples with focused rustdoc examples)
````
This shouldn't be here, it's a left over from the rebase.

luisschwab commented:
@kwsantiago you should run `just pre-push` to catch and correct any errors before pushing.

kwsantiago commented:
@luisschwab got it, thanks!

kwsantiago commented:
Code Coverage failure is unrelated to this PR: `test_chain_update_wrong_network_error` fails with `ConnectionRefused` on the local esplora server. Same failure on master: https://github.com/bitcoindevkit/bdk/actions/runs/22196832097

notmandatory commented:
This will need a rebase also since I merged #2123.

kwsantiago commented:
@notmandatory I rebased onto master in f9ec92156265591d8b140fbec54ddfa45328993f

ValuedMammal reviewed (changes requested):
- Can you change `ignore` to `no_run` in `bdk_electrum_client.rs`, that way we know the example compiles.
- You can remove the `#[allow(deprecated)]` from `fetch_latest_blocks`. Thanks @kwsantiago

ValuedMammal commented on crates/chain/src/tx_graph.rs:
```diff
@@ -645,6 +645,24 @@ impl<A: Anchor> TxGraph<A> {
+    ///     output: vec![],
+    /// };
+    ///
+    /// let changeset = graph.insert_tx(tx.clone());
```
nit: You could either comment out the empty tx, or demonstrate decoding a valid transaction in hex.

nit: Do we need to clone the tx before inserting it?

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -107,6 +116,26 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+    ///         0,
+    ///         constants::genesis_block(Network::Bitcoin).block_hash(),
+    ///     ))
+    ///     .build();
```
nit: It's common to use `spks_for_keychain` in the `FullScanRequest` as well.

```suggestion
    /// let request = FullScanRequest::<&str>::builder()
    ///     .spks_for_keychain("alice", [(0u32, script_pubkey)])
    ///     .chain_tip(CheckPoint::new(
    ///         0,
    ///         constants::genesis_block(Network::Bitcoin).block_hash(),
    ///     ))
    ///     .build();
```

ValuedMammal commented:
You may need to enable the crate default features for dev-dependencies in `electrum/Cargo.toml` in order for the doctests to compile.

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -190,6 +219,23 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+    ///     .spks([ScriptBuf::new_op_return(&[0x00; 20])])
+    ///     .build();
+    ///
+    /// let response = bdk_client.sync(request, 50, false)?;
```
nit: We normally recommend setting `fetch_prev_txouts: true`.

```suggestion
    /// let response = bdk_client.sync(request, 50, true)?;
```

ValuedMammal commented on crates/electrum/src/bdk_electrum_client.rs:
```diff
@@ -107,6 +116,26 @@ impl<E: ElectrumApi> BdkElectrumClient<E> {
+    ///     ))
+    ///     .build();
+    ///
+    /// let response = bdk_client.full_scan(request, 10, 50, false)?;
```
```suggestion
    /// let response = bdk_client.full_scan(request, 10, 50, true)?;
```

ValuedMammal commented:
In `cont_integration.yml` we no longer need to `--exclude 'example_*'` for the build-test job.

kwsantiago commented:
@ValuedMammal your comments have been addressed in 35d6980684e8d9f474dc6900e6853f700c09851b

nymius commented on crates/chain/src/indexer/keychain_txout.rs:
```diff
@@ -478,6 +478,22 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+    /// use bdk_chain::miniscript::{Descriptor, DescriptorPublicKey};
+    /// # use std::str::FromStr;
+    ///
+    /// let mut index = KeychainTxOutIndex::<&str>::new(10, true);
```
I would like to hear other's opinions, but wouldn't it be better to replace all these uses of raw integers as parameters by the name of the parameter itself, as we are talking of documentation?

notmandatory replied:
I agree, having the variable named after the parameter would be more readable than an unexplained value.

nymius reviewed (commented):
I think this is in a very good state, but It's hard to extract from the original issue the requirements to consider this done.

The value of the `examples/` was mainly due to the integration of the multiple components bdk implements, and that's hard to replicate by just adding doc examples to the interesting methods.
That's why I think there is no much escape to having to look at `bdk-cli` for it.

However, there are components missing examples in the docs that will benefit from these changes.

@kwsantiago has already implemented some. The following list is an attempt to extract the provided functionality by the `examples/` and identify the missing doc examples:

- [ ] `Esplora` new: `Blocking/AsyncClient::from_builder()`
> Missing example.
- [ ] `Esplora` sync: `EsploraAsync/EsploraExt::sync()`
> Missing example.
- [x] `Esplora` full scan: `EsploraAsync/EsploraExt::full_scan()` 
- [x] `Electrum` new: `BdkElectrumClient::new()`
- [x] `Electrum` sync: `BdkElectrumClient::sync()`
- [x] `Electrum` full scan: `BdkElectrumClient::full_scan()`
- [ ] `Electrum` parameter: `batch_size`.
- [ ] `bitcoind_rpc` sync/live: `Emitter::next_block()` and `Emitter::mempool`
> Missing example.
- [ ] `TxOut` listing: `TxGraph::try_canonical_view` ~~+ `filter_outpoints`~~   (`filter_outpoints`already contains examples)
> Missing example.
- [x] `Balance` `TxGraph::balance()`
- [x] `Index` setup: `KeychainTxOutIndex::insert_descriptor()`
- [ ] `Address` listing: `KeychainTxOutIndex::revealed_keychain_spks()`
- [x] `Address` reveal next: `KeychainTxOutIndex::reveal_next_spk()`
- [ ] `Address` reveal unused: `KeychainTxOutIndex::next_unused_spk()`

Once we have these in place, the added examples together with the `bdk-cli` project are going to be a better guide than the old `examples/`.

I left outside of the above list the functionality that cannot be replicated without recreating end to end examples like before.
- **Wallet setup**: generate keys, initialize indexes and local chain, set network and database. Trying to replicate this in some way will fall back again to the old `examples/` way. It's better to point users to `bdk-cli` instead.
- **`PSBT` operations**: new, sign and extract: PSBT creation is mainly managed through `TxBuilder` in `bdk_wallet`, `sign` is on its way to be removed and extraction is also done by `bdk_wallet`. For more fine grained control is better to look at `bdk_tx`.
- **Coin selection and transaction building**: Same than above, most of this is implemented with `TxBuilder` in `bdk_wallet`, and for the new interface is better to work on `bdk-tx`.

oleonardolima commented:
> I think this is in a very good state, but It's hard to extract from the original issue the requirements to consider this done.
> 
> The value of the `examples/` was mainly due to the integration of the multiple components bdk implements, and that's hard to replicate by just adding doc examples to the interesting methods. That's why I think there is no much escape to having to look at `bdk-cli` for it.
> 
> [...]

@nymius Thanks for the detailed comment and review! If I may add to it, I would suggest to break this into different PRs if possible, e.g one for bdk_esplora; one for bdk_electrum, and so on.

> I left outside of the above list the functionality that cannot be replicated without recreating end to end examples like before.
> 
> * **Wallet setup**: generate keys, initialize indexes and local chain, set network and database. Trying to replicate this in some way will fall back again to the old `examples/` way. It's better to point users to `bdk-cli` instead.
> * **`PSBT` operations**: new, sign and extract: PSBT creation is mainly managed through `TxBuilder` in `bdk_wallet`, `sign` is on its way to be removed and extraction is also done by `bdk_wallet`. For more fine grained control is better to look at `bdk_tx`.
> * **Coin selection and transaction building**: Same than above, most of this is implemented with `TxBuilder` in `bdk_wallet`, and for the new interface is better to work on `bdk-tx`.

Yes, it's fine to left this on the side for now, they're `bdk_wallet` concerns. And I even feel that some of the examples here and there are some sort of duplicates (needs double-checking).

notmandatory commented:
Probably better to do this as smaller focused PRs that are easier to review.

nymius commented:
@kwsantiago, could you split this PR in smaller focused changes? I wouldn't open the second until we have merged the first.

kwsantiago commented:
@nymius can you please provide a list of what the first and subsequent PRs would require from this PR, and I can start with the first one?

I've already had a lot of back and forth feedback with @oleonardolima over a 9 month period and I'd rather get precise instructions on what is needed to move this forward before I start any further work on this PR. Thank you.

nymius commented:
I cannot give you precise instructions, but guidance. The work you did is in great shape already. I left a review in a previous comment (https://github.com/bitcoindevkit/bdk/pull/2006#pullrequestreview-3932256491) that makes a list of the functionality that was not documented after the removal of the examples.

I would open a PR for removing a single example, and map the new documents added to the removed example. For example: remove `example_bitcoind_rpc_polling` and introduce the new documentation.

I would recommend you to leave `example_cli` to the end, as it is a dependency of most of the examples.

kwsantiago commented:
@nymius I opened #2210 to remove `example_bitcoind_rpc_polling` and add the Emitter docs; `esplora`, `electrum`, then `example_cli` will follow one at a time once it merges.

luisschwab commented:
@kwsantiago needs rebase

kwsantiago commented:
@luisschwab per @nymius this PR is being split into smaller ones (#2210 first, then esplora/electrum/example_cli) and closed out once they land, so I left it un-rebased; did you mean #2210?

luisschwab commented:
> this PR is being split into smaller ones

You can mark this PR as draft if that's the case.

kwsantiago committed:
chore: Remove example_cli

Remove the standalone example crates under examples/ and clean up
workspace, docs, and CI references that depended on them.

This commit only does the removal and cleanup. Adding targeted
rustdoc examples is deferred to follow-up PRs.

ValuedMammal commented:
Thanks @kwsantiago. Sorry for the delay.

As @nymius said, the rustdoc examples aren't a direct replacement for the example binaries, more of a nice to have, so it's best to treat this as two separate issues: first the removal of `example_cli` and dependent crates (this PR), then adding examples to the documentation as needed.

ACK 522e441a1a4511860a90c43c3932cd45a8797c15
